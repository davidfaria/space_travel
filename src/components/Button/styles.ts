import styled, { keyframes } from 'styled-components'

export const Wrapper = styled.button`
  position: relative;
  overflow: hidden;
  height: 44px;
  border-radius: 4px;
  border: 0;
  padding: 0 1rem;
  font-weight: 500;
  width: 100%;

  transition: Background-color 0.2s;

  &:hover:not(:disabled) {
    background-color: var(--primary-hover);
  }
  :disabled {
    cursor: not-allowed;
    opacity: 0.5;
  }

  &.primary {
    color: var(--button-text);
    background-color: var(--button-primary);
    border: 1px solid var(--button-primary);
  }
  &.primary:hover {
    color: var(--button-text);
    background-color: var(--button-primary-hover);
    border: 1px solid var(--button-primary-border);
  }

  &.outline-primary {
    color: var(--button-primary);
    background-color: transparent;
    border: 1px solid var(--button-primary);
  }
  &.outline-primary:hover {
    color: var(--button-text);
    background-color: var(--button-primary);
    border: 1px solid var(--button-primary);
  }
`

const rotate360 = keyframes`
  from {
    transform: rotate(0deg);
  }
  to {
    transform: rotate(360deg);
  }
`

export const Spinner = styled.div`
  animation: ${rotate360} 1s linear infinite;
  transform: translateZ(0);

  border-top: 2px solid #fff;
  border-right: 2px solid #fff;
  border-bottom: 2px solid #fff;
  border-left: 4px solid #007bff;
  background: transparent;
  width: 24px;
  height: 24px;
  border-radius: 50%;

  &.loader {
    margin: 0 auto;
  }
`
