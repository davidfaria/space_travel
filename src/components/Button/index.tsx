import React, { ButtonHTMLAttributes } from 'react'
import { IconBaseProps } from 'react-icons'
import * as S from './styles'

type Props = ButtonHTMLAttributes<HTMLButtonElement> & {
  icon?: React.ComponentType<IconBaseProps>
  loading?: boolean
  variant?: 'primary'
}

interface DefaultProps {
  variant?: string
}

export type ButtonProps = Props & DefaultProps

const Button: React.FC<ButtonProps> = ({
  icon: Icon,
  children,
  loading,
  variant = 'primary',
  ...rest
}) => {
  return (
    <S.Wrapper type="button" className={variant} {...rest}>
      {Icon && <Icon data-testid="button-icon" size={20} />}
      {loading ? (
        <S.Spinner data-testid="loader" className="loader" />
      ) : (
        children
      )}
    </S.Wrapper>
  )
}

export default Button
