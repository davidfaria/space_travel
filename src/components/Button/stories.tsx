import { Story, Meta } from '@storybook/react/types-6-0'
import { MdSearch } from 'react-icons/md'
import Button, { ButtonProps } from '.'

export default {
  title: 'Button',
  component: Button
} as Meta

export const Default: Story<ButtonProps> = (args) => (
  <Button {...args} icon={MdSearch}>
    Pesquisar
  </Button>
)
export const LoadingButton: Story<ButtonProps> = (args) => (
  <Button {...args} loading={true}>
    Pesquisar
  </Button>
)
