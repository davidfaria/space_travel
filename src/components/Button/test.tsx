import { render, screen } from '@testing-library/react'
import { MdSearch } from 'react-icons/md'

import Button from '.'

describe('<Button />', () => {
  it('should render the button', () => {
    const { container } = render(<Button>Pesquisar</Button>)

    expect(screen.getByRole('button')).toBeInTheDocument()
    expect(screen.getByText(/pesquisar/i)).toBeInTheDocument()

    expect(container.firstChild).toMatchSnapshot()
  })

  it('should render spinner when the button loading is true', () => {
    const { container } = render(<Button loading={true}>Pesquisar</Button>)

    expect(screen.getByTestId('loader')).toHaveClass('loader')
    expect(screen.queryByText(/pesquisar/i)).not.toBeInTheDocument()

    expect(container.firstChild).toMatchSnapshot()
  })

  it('should render button with icon', () => {
    const { container } = render(<Button icon={MdSearch}>Pesquisar</Button>)

    expect(screen.getByRole('button')).toBeInTheDocument()
    expect(screen.getByTestId('button-icon')).toBeInTheDocument()

    expect(container.firstChild).toMatchSnapshot()
  })
})
