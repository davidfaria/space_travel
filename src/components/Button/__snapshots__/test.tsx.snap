// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Button /> should render button with icon 1`] = `
.c0 {
  position: relative;
  overflow: hidden;
  height: 44px;
  border-radius: 4px;
  border: 0;
  padding: 0 1rem;
  font-weight: 500;
  width: 100%;
  -webkit-transition: Background-color 0.2s;
  transition: Background-color 0.2s;
}

.c0:hover:not(:disabled) {
  background-color: var(--primary-hover);
}

.c0:disabled {
  cursor: not-allowed;
  opacity: 0.5;
}

.c0.primary {
  color: var(--button-text);
  background-color: var(--button-primary);
  border: 1px solid var(--button-primary);
}

.c0.primary:hover {
  color: var(--button-text);
  background-color: var(--button-primary-hover);
  border: 1px solid var(--button-primary-border);
}

.c0.outline-primary {
  color: var(--button-primary);
  background-color: transparent;
  border: 1px solid var(--button-primary);
}

.c0.outline-primary:hover {
  color: var(--button-text);
  background-color: var(--button-primary);
  border: 1px solid var(--button-primary);
}

<button
  class="c0 primary"
  type="button"
>
  <svg
    data-testid="button-icon"
    fill="currentColor"
    height="20"
    stroke="currentColor"
    stroke-width="0"
    viewBox="0 0 24 24"
    width="20"
    xmlns="http://www.w3.org/2000/svg"
  >
    <path
      d="M15.5 14h-.79l-.28-.27C15.41 12.59 16 11.11 16 9.5 16 5.91 13.09 3 9.5 3S3 5.91 3 9.5 5.91 16 9.5 16c1.61 0 3.09-.59 4.23-1.57l.27.28v.79l5 4.99L20.49 19l-4.99-5zm-6 0C7.01 14 5 11.99 5 9.5S7.01 5 9.5 5 14 7.01 14 9.5 11.99 14 9.5 14z"
    />
  </svg>
  Pesquisar
</button>
`

exports[
  `<Button /> should render spinner when the button loading is true 1`
] = `
.c0 {
  position: relative;
  overflow: hidden;
  height: 44px;
  border-radius: 4px;
  border: 0;
  padding: 0 1rem;
  font-weight: 500;
  width: 100%;
  -webkit-transition: Background-color 0.2s;
  transition: Background-color 0.2s;
}

.c0:hover:not(:disabled) {
  background-color: var(--primary-hover);
}

.c0:disabled {
  cursor: not-allowed;
  opacity: 0.5;
}

.c0.primary {
  color: var(--button-text);
  background-color: var(--button-primary);
  border: 1px solid var(--button-primary);
}

.c0.primary:hover {
  color: var(--button-text);
  background-color: var(--button-primary-hover);
  border: 1px solid var(--button-primary-border);
}

.c0.outline-primary {
  color: var(--button-primary);
  background-color: transparent;
  border: 1px solid var(--button-primary);
}

.c0.outline-primary:hover {
  color: var(--button-text);
  background-color: var(--button-primary);
  border: 1px solid var(--button-primary);
}

.c1 {
  -webkit-animation: eoUyJr 1s linear infinite;
  animation: eoUyJr 1s linear infinite;
  -webkit-transform: translateZ(0);
  -ms-transform: translateZ(0);
  transform: translateZ(0);
  border-top: 2px solid #fff;
  border-right: 2px solid #fff;
  border-bottom: 2px solid #fff;
  border-left: 4px solid #007bff;
  background: transparent;
  width: 24px;
  height: 24px;
  border-radius: 50%;
}

.c1.loader {
  margin: 0 auto;
}

<button
  class="c0 primary"
  type="button"
>
  <div
    class="c1 loader"
    data-testid="loader"
  />
</button>
`

exports[`<Button /> should render the button 1`] = `
.c0 {
  position: relative;
  overflow: hidden;
  height: 44px;
  border-radius: 4px;
  border: 0;
  padding: 0 1rem;
  font-weight: 500;
  width: 100%;
  -webkit-transition: Background-color 0.2s;
  transition: Background-color 0.2s;
}

.c0:hover:not(:disabled) {
  background-color: var(--primary-hover);
}

.c0:disabled {
  cursor: not-allowed;
  opacity: 0.5;
}

.c0.primary {
  color: var(--button-text);
  background-color: var(--button-primary);
  border: 1px solid var(--button-primary);
}

.c0.primary:hover {
  color: var(--button-text);
  background-color: var(--button-primary-hover);
  border: 1px solid var(--button-primary-border);
}

.c0.outline-primary {
  color: var(--button-primary);
  background-color: transparent;
  border: 1px solid var(--button-primary);
}

.c0.outline-primary:hover {
  color: var(--button-text);
  background-color: var(--button-primary);
  border: 1px solid var(--button-primary);
}

<button
  class="c0 primary"
  type="button"
>
  Pesquisar
</button>
`
